WPT + GH Actions (#346)
oddbird/oddleventy: content/blog/2023/web-platform-tests-polyfills.md

---
title: Using Web Platform Tests to Write and Test Polyfills
author: ed
date: 2023-04-20
tags:
  - Article
  - Build Tools
  - DevOps
  - GitHub
  - JavaScript
  - Testing
image:
  src: blog/2023/graphs-and-charts.png
summary: |

>   Learn how to leverage Web Platform Tests to give your polyfills the confidence

Reviewer: ```suggestion
  Learn how to leverage Web Platform Tests to ensure your polyfills are implementing upcoming browser features correctly,
```

>   that they are implementing upcoming browser features correctly, including a

Reviewer: ```suggestion
  including how to generate a
```

serve` command. The tests are organized in folders, and visiting any of the HTML
files will run all subtests and display the results.

To get a bird's eye view of the test results for upcoming versions of major
browsers, visit [wpt.fyi](https://wpt.fyi).

## WPT and Polyfills

The primary users of WPT are browser vendors who want to evaluate development
versions of their browser against the test suite. However, polyfill authors can
also leverage WPT to run the test suite on stable browsers with their polyfill
loaded. This is useful to measure the impact of the polyfill in a quantitative
way that doesn't rely on manual testing.

WPT includes built-in support to run with an arbitrary JavaScript file inserted
as part of the server response. The process is explained in the [official
documentation](https://web-platform-tests.org/running-tests/testing-polyfills.html),
but it boils down to these steps:

1. Package your polyfill as a single-file, immediately-invoked script
2. Start WPT with `./wpt serve --inject-script=/path/to/polyfill.js`
3. Visit any WPT and ensure the script is inlined as the first element of the
   `<head>`. This should be obvious when inspecting the page source.

> At this point you should hopefully notice test scores have improved with the

Reviewer: Maybe we could add a screenshot to show an example of what those score results look like?

help of the injected polyfill. If something goes wrong, the WPT should also
report tests that are not completing or are failing for some other reason.

## Monitoring WPT "Coverage"

An important part of polyfill development is ensuring the polyfill covers all or

> most of the WPT for a given feature to maintain compliance with the spec.

Reviewer: This might be a place to swap `WPT` with `tests` since we're referring to the tests here and not the wider WPT project

---
title: Using Web Platform Tests to Write and Test Polyfills
author: ed

> date: 2023-04-29

Reviewer 2: Maybe a Monday publish date?

You can run all tests in your current browser by visiting
[wpt.live](https://wpt.live) or by cloning the repository and running the `./wpt
serve` command. The tests are organized in folders, and visiting any of the HTML
files will run all subtests and display the results. For example, visiting the
test
[`anchor-name-0002.html`](https://wpt.live/css/css-anchor-position/anchor-name-002.html)
with Chrome 112 results in two passing and two failing tests:

> {{ embed.img(
>   src='blog/2023/wpt-test-fail.png',
>   alt='Test results for anchor-name-0002.html, showing two of four passing tests'
> ) }}

Reviewer 3: I think if we have a larger image, that would fill the space and we can use a less pixelated image perhaps.

Something like this one might work to full the column:  
![wpt-tests-fail](https://user-images.githubusercontent.com/1581694/234417328-57ddc8a5-1f15-430e-a1f2-2b4f96fd1f34.jpg)

Either that or if we want the smaller image, we do have some alignment classes:

```suggestion
{{ embed.img(
  src='blog/2023/wpt-test-fail.png',
  alt='Test results for anchor-name-0002.html, showing two of four passing tests',
  attrs={'class': 'align-center'}
) }}
```

We used to have an img-border class but I don't think any CSS exists for that anymore. I will make a card to either add the CSS back in for that or remove the classes in the markup. I don't think the extend-left and extend-right work anymore, at least not on these images.

> ) }}

Author: I used larger images in ae7c77d81fa6f02170c7775225c5f3ed092596ef. Now they cover the entire column
